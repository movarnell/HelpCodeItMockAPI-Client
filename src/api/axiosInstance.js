// src/api/axiosInstance.js
import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'http://localhost:5100', // Adjust if necessary
});

axiosInstance.interceptors.request.use(
    (config) => {
      const token = localStorage.getItem('token');
      console.log('Token from localStorage:', token);
      if (token) {
        config.headers['Authorization'] = `Bearer ${token}`;
      } else {
        console.warn('No token found in localStorage');
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

export default axiosInstance;