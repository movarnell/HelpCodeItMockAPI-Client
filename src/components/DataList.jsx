// src/components/DataList.jsx
import React, { useEffect, useState } from 'react';
import axiosInstance from '../api/axiosInstance';
import { useParams, Link } from 'react-router-dom';
import { IoArrowBack } from 'react-icons/io5';
function DataList() {
  const { endpointName } = useParams();
  const [dataList, setDataList] = useState([]);
  const [error, setError] = useState('');

  // Fetch data from the backend
  const fetchData = async () => {
    try {
      const response = await axiosInstance.get(`/api/${endpointName}`);
      console.log('Fetched Data:', response.data); // Debugging line
      setDataList(response.data);
    } catch (err) {
      console.error(err);
      setError('Failed to fetch data.');
    }
  };

  useEffect(() => {
    fetchData();
  }, [endpointName]);

  // Handle delete action
  const handleDelete = async (dataId) => {
    if (!window.confirm('Are you sure you want to delete this entry?')) {
      return;
    }

    console.log(`Attempting to delete data_id: ${dataId}`); // Debugging line

    try {
      await axiosInstance.delete(`/api/${endpointName}`, {
        params: { id: dataId }, // Using params option
      });
      console.log(`Deleted data_id: ${dataId}`); // Debugging line
      // Update the dataList state to remove the deleted item
      setDataList(dataList.filter(item => item.data_id !== dataId));
    } catch (err) {
      console.error(err);
      setError(err.response?.data?.message || 'Failed to delete data.');
    }
  };

  return (
    <div className="p-6">
      <div className="flex items-center mb-6">
        <Link
              to="/dashboard"
              className="mr-4 text-indigo-600 transition-colors duration-200 hover:text-indigo-800"
            >
              <IoArrowBack className="w-8 h-8" />
            </Link>
        <h1 className="text-3xl font-bold ">Data for the {endpointName} endpoint</h1>
      </div>
      {error && <p className="text-red-500">{error}</p>}
      <Link
        to={`/api/${endpointName}/add`}
        className="px-4 py-2 mb-4 font-bold text-white bg-blue-500 rounded-md hover:bg-blue-600"
      >
        Add Data
      </Link>
      <table className="w-full mt-10 text-left border border-collapse border-gray-300 table-auto">
        <thead>
          <tr className="bg-gray-200">
            {dataList.length > 0 &&
              Object.keys(dataList[0]).map((key) => (
                key !== 'data_id' && ( // Optionally hide data_id from the table
                  <th key={key} className="px-4 py-2 border border-gray-300">
                    {key}
                  </th>
                )
              ))}
            <th className="px-4 py-2 border border-gray-300">Actions</th>
          </tr>
        </thead>
        <tbody>
          {dataList.map((item) => (
            <tr key={item.data_id} className="hover:bg-gray-100">
              {Object.entries(item).map(([key, value]) => (
                key !== 'data_id' && ( // Optionally hide data_id from the table
                  <td key={key} className="px-4 py-2 border border-gray-300">
                    {value}
                  </td>
                )
              ))}
              <td className="px-4 py-2 border border-gray-300">
                <button
                  onClick={() => handleDelete(item.data_id)}
                  className="px-2 py-1 text-sm text-white bg-red-500 rounded hover:bg-red-600"
                >
                  Delete
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
      <div className="px-4 py-2 mt-24 mb-4 font-bold text-white bg-green-500 rounded-md hover:bg-green-600 w-fit">
        <Link to={'/dashboard'}> Back to Dashboard</Link>
      </div>
    </div>
  );
}

export default DataList;