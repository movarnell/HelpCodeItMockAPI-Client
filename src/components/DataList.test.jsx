import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter, useParams } from 'react-router-dom';
import DataList from './DataList';
import axiosInstance from '../api/axiosInstance';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';

jest.mock('../api/axiosInstance');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  useParams: jest.fn(),
}));

describe('DataList Component', () => {
  const mockData = [
    { id: 1, name: 'Test 1' },
    { id: 2, name: 'Test 2' },
  ];

  beforeEach(() => {
    useParams.mockReturnValue({ endpointName: 'test_endpoint' });
    axiosInstance.get.mockResolvedValue({ data: mockData });
  });

  const renderComponent = async () => {
    await act(async () => {
      render(
        <BrowserRouter>
          <DataList />
        </BrowserRouter>
      );
    });
    await waitFor(() => expect(axiosInstance.get).toHaveBeenCalled());
  };

  test('renders data list', async () => {
    await renderComponent();
    expect(screen.getByText(/data for the test_endpoint endpoint/i)).toBeInTheDocument();
    expect(screen.getByText('Test 1')).toBeInTheDocument();
    expect(screen.getByText('Test 2')).toBeInTheDocument();
  });

  test('renders add data button', async () => {
    await renderComponent();
    expect(screen.getByText('Add Data')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <DataList />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});