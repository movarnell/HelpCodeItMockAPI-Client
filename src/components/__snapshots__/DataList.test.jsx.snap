// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList Component matches snapshot 1`] = `
<div
  className="p-6"
>
  <div
    className="flex items-center mb-6"
  >
    <a
      className="mr-4 text-indigo-600 transition-colors duration-200 hover:text-indigo-800"
      href="/dashboard"
      onClick={[Function]}
    >
      <svg
        className="w-8 h-8"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M244 400 100 256l144-144M120 256h292"
          fill="none"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="48"
        />
      </svg>
    </a>
    <h1
      className="text-3xl font-bold "
    >
      Data for the 
      test_endpoint
       endpoint
    </h1>
  </div>
  <a
    className="px-4 py-2 mb-4 font-bold text-white bg-blue-500 rounded-md hover:bg-blue-600"
    href="/api/test_endpoint/add"
    onClick={[Function]}
  >
    Add Data
  </a>
  <table
    className="w-full mt-10 text-left border border-collapse border-gray-300 table-auto"
  >
    <thead>
      <tr
        className="bg-gray-200"
      >
        <th
          className="px-4 py-2 border border-gray-300"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div
    className="px-4 py-2 mt-24 mb-4 font-bold text-white bg-green-500 rounded-md hover:bg-green-600 w-fit"
  >
    <a
      href="/dashboard"
      onClick={[Function]}
    >
       Back to Dashboard
    </a>
  </div>
</div>
`;
